src/Identity/Core/src/Passkeys/BufferSourceJsonConverter.cs

```diff
@@ -0,0 +1,99 @@
+        if (reader.ValueIsEscaped)
+        {
+            // We currently don't handle escaped base64url values, as we don't expect
+            // to encounter them when reading payloads produced by WebAuthn clients.
+            // See: https://www.w3.org/TR/webauthn-3/#base64url-encoding
+            throw new JsonException("Unexpected escaped value in base64url string.");
+        }
+
+        var span = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
+        if (!TryDecodeBase64Url(span, out var bytes))
+        {
+            throw new JsonException("Expected a valid base64url string.");
+        }
+
+        return BufferSource.FromBytes(bytes);
+    }
+
+    public override void Write(Utf8JsonWriter writer, BufferSource value, JsonSerializerOptions options)
+    {
+        var bytes = value.AsSpan();
+        WriteBase64UrlStringValue(writer, bytes);
+    }
+
+    // Based on https://github.com/dotnet/runtime/blob/624737eb3796e1a760465912b27ac349965d8ba5/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs#L218
+    private static bool TryDecodeBase64Url(ReadOnlySpan<byte> utf8Unescaped, [NotNullWhen(true)] out byte[]? bytes)
+    {
+        byte[]? pooledArray = null;
+
+        Span<byte> byteSpan = utf8Unescaped.Length <= StackallocByteThreshold ?
+            stackalloc byte[StackallocByteThreshold] :
+            (pooledArray = ArrayPool<byte>.Shared.Rent(utf8Unescaped.Length));
+
+        var status = Base64Url.DecodeFromUtf8(utf8Unescaped, byteSpan, out var bytesConsumed, out var bytesWritten);
+        if (status != OperationStatus.Done)
+        {
+            bytes = null;
+
+            if (pooledArray != null)
+            {
+                byteSpan.Clear();
+                ArrayPool<byte>.Shared.Return(pooledArray);
+            }
+
+            return false;
+        }
+        Debug.Assert(bytesConsumed == utf8Unescaped.Length);
+
+        bytes = byteSpan[..bytesWritten].ToArray();
+
+        if (pooledArray != null)
+        {
+            byteSpan.Clear();
+            ArrayPool<byte>.Shared.Return(pooledArray);
+        }
+
+        return true;
+    }
+
+    private static void WriteBase64UrlStringValue(Utf8JsonWriter writer, ReadOnlySpan<byte> bytes)
+    {
+        byte[]? pooledArray = null;
+
+        var encodedLength = Base64Url.GetEncodedLength(bytes.Length);
+        var byteSpan = encodedLength <= StackallocByteThreshold ?
+            stackalloc byte[encodedLength] :
+            (pooledArray = ArrayPool<byte>.Shared.Rent(encodedLength));
+
+        var status = Base64Url.EncodeToUtf8(bytes, byteSpan, out var bytesConsumed, out var bytesWritten);
+        Debug.Assert(status == OperationStatus.Done);
+        Debug.Assert(bytesConsumed == bytes.Length);
+
+        var base64UrlUtf8 = byteSpan[..bytesWritten];
+        writer.WriteStringValue(base64UrlUtf8);
+
+        if (pooledArray != null)
+        {
+            byteSpan.Clear();
```

Comment:
```suggestion
```

If we really cared about clearing we could just call `ArrayPool<byte>.Shared.Return(pooledArray, clearArray: true);` one line below, but I don't think it's necessary. We don't typically clear the pooled memory we write request and response data into either. I think we can get rid of the other calls to Clear() as well.